fetchRebalancerData.js

```diff
@@ -9,68 +11,94 @@ const contractAbi = require('./UniswapNFTPositionManagerAbi.json');
-
-    await contract.getPastEvents("IncreaseLiquidity",
-        { fromBlock: 25698992, toBlock: 25774738 },
-        (err, events) => {
-            for (let i = 0; i < events.length; i++) {
-                // console.log(events[i].transactionHash);
-                transactionHashes.add(events[i].transactionHash);
-            }
-            // console.log(events[0]);
-            // transactionHash = events[0].transactionHash;
-        });
-
-    events = await contract.getPastEvents("DecreaseLiquidity",
-        { fromBlock: 25698992, toBlock: 25774738 }
-    );
-
-    // console.log(events);
-    const instaTransactionHashes = events.map(e => {
-        return {
-            transactionHash: e.transactionHash,
-            tokenId: e.returnValues.tokenId,
-            liquidity: e.returnValues.liquidity
-        };
-    }).filter(transaction => transactionHashes.has(transaction.transactionHash));
-    // console.log(instaTransactionHashes);
-
-    finalTransactions = await Promise.all(instaTransactionHashes.map(async (transaction) => {
+let transactionData;
+
+const getEvents = async (dsaAccounts) => {
+    // store all transaction hashes obtained from IncreaseLiquidity event emits in a set for quick search later on
+    const transactionHashes = new Map();
+
+    let latestBlock = await web3.eth.getBlockNumber(); // get the latest block number
+    let allValidTransactions = [];
+    let currentToBlock;
+
+    for (let currentFirstBlock = 25657968; currentFirstBlock < latestBlock; currentFirstBlock = currentToBlock + 1) {
+        currentToBlock = Math.min(latestBlock, currentFirstBlock + 3000);
+        // get past events of type IncreaseLiquidity between currentFirstBlock and currentToBlock block numbers from Uniswap contract
+        await contract.getPastEvents("IncreaseLiquidity",
+            { fromBlock: currentFirstBlock, toBlock: currentToBlock },
+            (err, events) => {
+                // push the transaction hash of all emitted events into the set
+                for (let i = 0; i < events.length; i++) {
+                    transactionHashes.set(events[i].transactionHash, events[i].returnValues.tokenId);
+                }
+            });
+
+        // get past events of type DecreaseLiquidity between currentFirstBlock and currentToBlock block numbers
+        events = await contract.getPastEvents("DecreaseLiquidity",
+            { fromBlock: currentFirstBlock, toBlock: currentToBlock }
+        );
+
+        // filter out transaction hashes between the current block range that emitted both IncreaseLiquidity and DecreaseLiquidity events
+        // also filter according to the condition that the tokenIds for IncreaseLiquidity and DecreaseLiquidity events must be different
+        const validTransactionInRange = events
+            .filter(transaction => (transactionHashes.has(transaction.transactionHash) && transaction.transactionHash !== transactionHashes.get(transaction.transactionHash)))
```

Comment:
there should also be this condition:
`newTokenId != oldTokenID`.

Reply:
Oh right, put `.transactionHash` instead of `.returnValues.tokenId` by mistake, fixed.